File: tests/encoding_test.cpp
// The move encoding's perft (docs/01-engine.md): for every legal move in
// every position of a depth-4 walk from the five perft suite positions,
// index_to_move(move_to_index(m)) == m and no two legal moves in a position
// share an index. Plus a handful of hand-computed convention pins so a silent
// change of direction ordering reads as "e2e4 is no longer 877", not as a
// mysteriously retrained-from-scratch net.
//
// The plane encoder is deliberately not tested here: its oracle is the pure
// Python reference in tests/encoding_test.py, which shares no code with the
// C++ encoder.
#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#include "encoding.hpp"

namespace {

constexpr const char* kSuite[] = {
    "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
    "r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R w KQkq - 0 1",
    "8/2p5/3p4/KP5r/1R3p1k/8/4P1P1/8 w - - 0 1",
    "r3k2r/Pppp1ppp/1b3nbN/nP6/BBP1P3/q4N2/Pp1P2PP/R2Q1RK1 w kq - 0 1",
    "rnbq1k1r/pp1Pbppp/2p5/8/2B5/8/PPP1NnPP/RNBQK2R w KQ - 0 1",
};

std::atomic<std::uint64_t> g_failures{0};
std::atomic<std::uint64_t> g_positions{0};
std::atomic<std::uint64_t> g_moves{0};
std::mutex g_print_mutex;

void fail(const std::string& what) {
    // Cap the noise: the count is exact, the messages are diagnostics.
    if (g_failures.fetch_add(1) < 20) {
        std::lock_guard<std::mutex> lock(g_print_mutex);
        std::cout << "  FAIL: " << what << "\n";
    }
}

// Checks one position and recurses. Collision detection via sorted indices;
// round-trip through the real az::index_to_move, movelist regeneration and
// all — this is the exposed function under test, not a shortcut of it.
void walk(chess::Board& board, int depth) {
    chess::Movelist moves;
    chess::movegen::legalmoves(moves, board);

    g_positions.fetch_add(1, std::memory_order_relaxed);
    g_moves.fetch_add(moves.size(), std::memory_order_relaxed);

    int indices[256];
    for (int i = 0; i < moves.size(); ++i) {
        const int idx = az::move_to_index(board, moves[i]);
        indices[i]    = idx;
        if (idx < 0 || idx >= az::kPolicySize) {
            fail("index " + std::to_string(idx) + " out of range for " +
                 chess::uci::moveToUci(moves[i]) + " in " + board.getFen());
            continue;
        }
        if (az::index_to_move(board, idx) != moves[i]) {
            fail("round trip broke for " + chess::uci::moveToUci(moves[i]) + " (index " +
                 std::to_string(idx) + ") in " + board.getFen());
        }
    }

    std::sort(indices, indices + moves.size());
    if (std::adjacent_find(indices, indices + moves.size()) != indices + moves.size()) {
        fail("index collision in " + board.getFen());
    }

    if (depth <= 0) return;
    for (const auto& move : moves) {
        board.makeMove(move);
        walk(board, depth - 1);
        board.unmakeMove(move);
    }
}

// Hand-computed pins of the encoding convention (see encoding.hpp header
// comment for the tables these numbers come from).
void pin(const std::string& fen, const std::string& uci, int want) {
    chess::Board board;
    board.setFen(fen);
    chess::Movelist moves;
    chess::movegen::legalmoves(moves, board);
    for (const auto& move : moves) {
        if (chess::uci::moveToUci(move) == uci) {
            const int got = az::move_to_index(board, move);
            if (got != want) {
                fail("pin " + uci + " in " + fen + ": got " + std::to_string(got) + ", want " +
                     std::to_string(want));
            }
            return;
        }
    }
    fail("pin move " + uci + " is not legal in " + fen);
}

void convention_pins() {
    const std::string startpos = kSuite[0];
    const std::string castlepos = "r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1";
    const std::string castlepos_b = "r3k2r/8/8/8/8/8/8/R3K2R b KQkq - 0 1";

    pin(startpos, "e2e4", 12 * 73 + 1);            // N, distance 2          = 877
    pin(startpos, "g1f3", 6 * 73 + 63);            // knight (-1,+2)         = 501
    pin(castlepos, "e1g1", 4 * 73 + 2 * 7 + 1);    // king E, distance 2     = 307
    pin(castlepos, "e1c1", 4 * 73 + 6 * 7 + 1);    // king W, distance 2     = 335
    pin(castlepos_b, "e8g8", 4 * 73 + 2 * 7 + 1);  // same square+type once oriented
    pin("8/P6k/8/8/8/8/7K/8 w - - 0 1", "a7a8n", 48 * 73 + 64);      // fwd underpromo N = 3568
    pin("1n5k/P7/8/8/8/8/7K/8 w - - 0 1", "a7b8r", 48 * 73 + 72);    // capture-right underpromo R
    pin("7k/8/8/8/8/8/p6K/8 b - - 0 1", "a2a1n", 48 * 73 + 64);      // black mirrors to 3568
    pin("7k/4P3/8/8/8/8/8/K7 w - - 0 1", "e7e8q", 52 * 73 + 0);      // queen promo = plain N move
}

}  // namespace

int main(int argc, char** argv) {
    int depth   = 4;
    int threads = static_cast<int>(std::thread::hardware_concurrency());
    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--depth" && i + 1 < argc) depth = std::stoi(argv[++i]);
        if (arg == "--threads" && i + 1 < argc) threads = std::stoi(argv[++i]);
    }

    convention_pins();

    // Root-split: one job per (suite position, root move) subtree, same
    // pattern as perft_threaded. The five root positions themselves are
    // covered by walk(root, 0) first.
    std::vector<chess::Board> jobs;
    for (const char* fen : kSuite) {
        chess::Board root;
        root.setFen(fen);
        walk(root, 0);

        chess::Movelist moves;
        chess::movegen::legalmoves(moves, root);
        for (const auto& move : moves) {
            root.makeMove(move);
            jobs.push_back(root);
            root.unmakeMove(move);
        }
    }

    const auto t0 = std::chrono::steady_clock::now();
    std::atomic<std::size_t> next{0};
    std::vector<std::thread> pool;
    const int n_workers = std::max(1, std::min<int>(threads, static_cast<int>(jobs.size())));
    pool.reserve(n_workers);
    for (int t = 0; t < n_workers; ++t) {
        pool.emplace_back([&] {
            for (std::size_t i = next.fetch_add(1); i < jobs.size(); i = next.fetch_add(1)) {
                chess::Board board = jobs[i];
                walk(board, depth - 1);  // the root move consumed one ply
            }
        });
    }
    for (auto& th : pool) th.join();
    const auto elapsed = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();

    std::cout << "encoding walk: depth " << depth << ", " << g_positions.load() << " positions, "
              << g_moves.load() << " moves round-tripped in " << elapsed << "s\n";

    if (g_failures.load() != 0) {
        std::cout << "ENCODING TEST FAILED: " << g_failures.load() << " failure(s)\n";
        return 1;
    }
    std::cout << "encoding test OK\n";
    return 0;
}

File: tests/encoding_test.py
#!/usr/bin/env python
"""The plane encoder's oracle tests (docs/01-engine.md).

The C++ move round-trip walk (tests/encoding_test.cpp) is blind to plane bugs,
so the planes get two tests of their own here, with python-chess as the
independent oracle:

1. Reference equivalence: a pure-Python encoder written straight from the
   plane table in docs/01-engine.md, sharing no code with the C++ one, must
   match encode_batch byte-for-byte on every position of a depth-3 walk from
   the five perft suite positions, plus hand-picked FENs covering all 16
   castling-rights combinations, en passant on each file (both colors), and
   halfmove clocks {0, 49, 99, 100, 150}.

2. Color-mirror invariance: encode(P) == encode(mirror(P)) byte-for-byte with
   python-chess Board.mirror() as the oracle, and for every legal move m,
   move_to_index(m) == move_to_index(mirror(m)). This is the only test forcing
   "the mover's perspective" to mean the same flip in both encoders.

Run:  python tests/encoding_test.py [--depth 3]
"""

import argparse
import os
import sys

import chess
import numpy as np

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import chess_engine as ce  # the built module at the repo root

SUITE = [
    "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
    "r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R w KQkq - 0 1",
    "8/2p5/3p4/KP5r/1R3p1k/8/4P1P1/8 w - - 0 1",
    "r3k2r/Pppp1ppp/1b3nbN/nP6/BBP1P3/q4N2/Pp1P2PP/R2Q1RK1 w kq - 0 1",
    "rnbq1k1r/pp1Pbppp/2p5/8/2B5/8/PPP1NnPP/RNBQK2R w KQ - 0 1",
]

failures = 0


def fail(what):
    global failures
    failures += 1
    if failures <= 20:
        print(f"  FAIL: {what}")


# --- the ~30-line reference encoder, straight from the table -----------------

def encode_ref(fen):
    board = chess.Board(fen)
    stm = board.turn
    planes = np.zeros((19, 8, 8), dtype=np.uint8)
    for sq, piece in board.piece_map().items():
        if stm == chess.BLACK:
            sq = chess.square_mirror(sq)  # flip ranks, keep files
        base = 0 if piece.color == stm else 6
        planes[base + piece.piece_type - 1, sq // 8, sq % 8] = 1
    planes[12] = 1
    planes[13] = board.has_kingside_castling_rights(stm)
    planes[14] = board.has_queenside_castling_rights(stm)
    planes[15] = board.has_kingside_castling_rights(not stm)
    planes[16] = board.has_queenside_castling_rights(not stm)
    if board.has_legal_en_passant():
        planes[17, :, chess.square_file(board.ep_square)] = 1
    planes[18] = min(board.halfmove_clock, 100)
    return planes


# --- position generation ------------------------------------------------------

def fen_of(board):
    # "legal" en-passant semantics match the movegen library, which keeps the
    # ep square only when an en-passant capture is actually legal.
    return board.fen(en_passant="legal")


def collect_walk(root_fen, depth, out):
    board = chess.Board(root_fen)

    def rec(d):
        out.add(fen_of(board))
        if d == 0:
            return
        for move in board.legal_moves:
            board.push(move)
            rec(d - 1)
            board.pop()

    rec(depth)


def hand_picked():
    fens = []
    # All 16 castling-rights combinations, both sides to move.
    for stm in "wb":
        for i in range(16):
            rights = "".join(c for c, bit in zip("KQkq", (8, 4, 2, 1)) if i & bit) or "-"
            fens.append(f"r3k2r/8/8/8/8/8/8/R3K2R {stm} {rights} - 0 1")

    # A legal en-passant capture on each file, for both capturing colors.
    def row(files_to_chars):
        out, empty = "", 0
        for f in range(8):
            if f in files_to_chars:
                out += str(empty) if empty else ""
                out += files_to_chars[f]
                empty = 0
            else:
                empty += 1
        return out + (str(empty) if empty else "")

    for f in range(8):
        adj = f - 1 if f > 0 else f + 1
        sq = "abcdefgh"[f]
        fens.append(f"4k3/8/8/8/{row({f: 'P', adj: 'p'})}/8/8/4K3 b - {sq}3 0 1")
        fens.append(f"4k3/8/8/{row({f: 'p', adj: 'P'})}/8/8/8/4K3 w - {sq}6 0 1")

    # Halfmove clocks, including the >100 clamp.
    for clock in (0, 49, 99, 100, 150):
        fens.append(f"r3k2r/8/8/8/8/8/8/R3K2R w KQkq - {clock} 1")
    return fens


# --- mirror helpers -----------------------------------------------------------

def mirror_fen(fen):
    board = chess.Board(fen)
    mirrored = board.mirror()
    mirrored.halfmove_clock = board.halfmove_clock  # mirror() must not touch it
    mirrored.fullmove_number = board.fullmove_number
    return fen_of(mirrored)


_RANK_FLIP = str.maketrans("12345678", "87654321")


def mirror_uci(uci):
    return uci.translate(_RANK_FLIP)


# --- the tests ------------------------------------------------------------

def check_chunk(fens):
    cpp = np.asarray(ce.encode_batch(fens))
    ref = np.stack([encode_ref(f) for f in fens])
    for i in np.nonzero((cpp != ref).any(axis=(1, 2, 3)))[0]:
        plane = int(np.nonzero((cpp[i] != ref[i]).any(axis=(1, 2)))[0][0])
        fail(f"planes differ from reference at plane {plane} for {fens[i]}")

    mirrored = [mirror_fen(f) for f in fens]
    cpp_m = np.asarray(ce.encode_batch(mirrored))
    for i in np.nonzero((cpp != cpp_m).any(axis=(1, 2, 3)))[0]:
        fail(f"encode(P) != encode(mirror(P)) for {fens[i]} vs {mirrored[i]}")

    for fen, mfen in zip(fens, mirrored):
        moves, idxs = ce.legal_moves(fen), ce.legal_move_indices(fen)
        if len(set(idxs)) != len(moves):
            fail(f"move index collision in {fen}")
        m_index = dict(zip(ce.legal_moves(mfen), ce.legal_move_indices(mfen)))
        for uci, idx in zip(moves, idxs):
            if m_index.get(mirror_uci(uci)) != idx:
                fail(f"move_to_index({uci}) not mirror-invariant in {fen}")


def convention_pins():
    castle = "r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1"
    for fen, uci, want in [
        (ce.STARTPOS, "e2e4", 877),
        (ce.STARTPOS, "g1f3", 501),
        (castle, "e1g1", 307),
        (castle, "e1c1", 335),
        ("8/P6k/8/8/8/8/7K/8 w - - 0 1", "a7a8n", 3568),
    ]:
        got = ce.move_to_index(fen, uci)
        if got != want:
            fail(f"convention pin {uci}: got {got}, want {want}")
        if ce.index_to_move(fen, want) != uci:
            fail(f"index_to_move({want}) != {uci}")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--depth", type=int, default=3, help="walk depth from each suite position")
    parser.add_argument("--chunk", type=int, default=4096)
    args = parser.parse_args()

    assert ce.POLICY_SIZE == 4672 and ce.NUM_PLANES == 19
    convention_pins()

    game = ce.Game()
    if not np.array_equal(*np.asarray(ce.encode_batch([game, game.fen]))):
        fail("encode_batch(Game) != encode_batch(fen)")

    positions = set()
    for fen in SUITE:
        collect_walk(fen, args.depth, positions)
    positions.update(hand_picked())
    fens = sorted(positions)
    print(f"checking {len(fens)} positions (depth-{args.depth} walk + hand-picked)")

    for start in range(0, len(fens), args.chunk):
        check_chunk(fens[start:start + args.chunk])
        done = min(start + args.chunk, len(fens))
        if done % (args.chunk * 8) == 0 or done == len(fens):
            print(f"  {done}/{len(fens)}")

    if failures:
        print(f"ENCODING TEST FAILED: {failures} failure(s)")
        return 1
    print("encoding test OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())
